jetbrains-core/tst/software/aws/toolkits/jetbrains/services/cwc/auth/AuthControllerTest.kt

```diff
@@ -0,0 +1,185 @@
+
+        // Act
+        val authController = AuthController()
+        val authNeededState = authController.getAuthNeededState(mockProject)
+
+        // Assert
+        assertThat(authNeededState).isNotNull
+        assertThat(authNeededState?.authType).isEqualTo(AuthFollowUpType.FullAuth)
+    }
+
+    @Test
+    fun `NotConnected with connected CodeWhisperer returns MissingScope`() {
+        // Arrange
+        every { checkBearerConnectionValidity(any(), BearerTokenFeatureSet.Q) } returns ActiveConnection.NotConnected
+        every { checkBearerConnectionValidity(any(), BearerTokenFeatureSet.CODEWHISPERER) } returns mockk<ActiveConnection.ValidBearer>(relaxed = true)
+
+        // Act
+        val authController = AuthController()
+        val authNeededState = authController.getAuthNeededState(mockProject)
+
+        // Assert
+        assertThat(authNeededState).isNotNull
+        assertThat(authNeededState?.authType).isEqualTo(AuthFollowUpType.MissingScopes)
+    }
+
+    @Test
+    fun `ValidBearer returns null`() {
+        // Arrange
+        every { checkBearerConnectionValidity(any(), BearerTokenFeatureSet.Q) } returns mockk<ActiveConnection.ValidBearer>(relaxed = true)
+        every { checkBearerConnectionValidity(any(), BearerTokenFeatureSet.CODEWHISPERER) } returns ActiveConnection.NotConnected
+
+        // Act
+        val authController = AuthController()
+        val authNeededState = authController.getAuthNeededState(mockProject)
+
+        // Assert
+        assertThat(authNeededState).isNull()
+    }
+
+    @Test
+    fun `ExpiredBearer returns AuthNeededState`() {
+        // Arrange
+        every { checkBearerConnectionValidity(any(), BearerTokenFeatureSet.Q) } returns mockk<ActiveConnection.ExpiredBearer>(relaxed = true)
+        every { checkBearerConnectionValidity(any(), BearerTokenFeatureSet.CODEWHISPERER) } returns ActiveConnection.NotConnected
+
+        // Act
+        val authController = AuthController()
+        val authNeededState = authController.getAuthNeededState(mockProject)
+
+        // Assert
+        assertThat(authNeededState).isNotNull
+        assertThat(authNeededState?.authType).isEqualTo(AuthFollowUpType.ReAuth)
+    }
+
+    @Test
+    fun `ExpiredIam returns AuthNeededState`() {
+        // Arrange
+        every { checkBearerConnectionValidity(any(), BearerTokenFeatureSet.Q) } returns mockk<ActiveConnection.ExpiredIam>(relaxed = true)
+        every { checkBearerConnectionValidity(any(), BearerTokenFeatureSet.CODEWHISPERER) } returns ActiveConnection.NotConnected
+
+        // Act
+        val authController = AuthController()
+        val authNeededState = authController.getAuthNeededState(mockProject)
+
+        // Assert
+        assertThat(authNeededState).isNotNull
+        assertThat(authNeededState?.authType).isEqualTo(AuthFollowUpType.FullAuth)
+    }
+
+    /*
+    // Broken due to Kotlin glitch preventing mocking of outside method "GettingStartedAuthUtilsKt"
```

Comment:
I am unable to mock this method that is called with default credentials.  Unsure if it's related to that or because it's in a different project scope.  Regardless this causes the 'real' method to continue being called, thereby breaking the unit test. 

`// Broken here: cannot mock outside method
        mockkStatic("software.aws.toolkits.jetbrains.core.gettingstarted.GettingStartedAuthUtilsKt")
        every {
            requestCredentialsForQ(
                any(), // project
                any(), // initialConnectionCount, using any() since we can't replicate default logic
                any(), // initialAuthConnections, using any() for the same reason
                any(), // isFirstInstance
                any(), // connectionInitiatedFromExplorer
                eq(true) // connectionInitiatedFromQChatPanel, with specific value
            )
        } returns true`

Reply:
```kt
mockkStatic(::requestCredentialsForQ)
```

is the invocation you want

Reply:
Unfortunately this doesn't work - are we sure it's even possible to mock these standalone static methods in mockito?  They do not have an associated class which makes this impossible.  From my understanding having standalone methods is not ideal either - I feel like we need to have a separate PR to refactor the `GettingStartedAuthUtils` to be a proper static class before being able to mock it in this test file.